### PR TITLE
Replace npm with pnpm

### PR DESCRIPTION
**Pull request by menghif**
## Issue This PR Addresses

Initial work for #1778

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description

This PR changes the package manager from `npm` to [`pnpm`](https://pnpm.io/). pnpm is faster and uses less disk space.

To try this locally:
- clone the telescope repo to your computer (do not use your current local copy of telescope)
- install pnpm: `npm install -g pnpm`
- run: `pnpm install`

## Checklist

- [x] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

**Comment by menghif** [comment]
I'm working on updated GitHub Actions for this.

File: .prettierignore
Comment (humphd) [inline .prettierignore:13]: This seems like something that should happen on `master` in a separate PR vs. here (this is an unrelated change).  If w need this, can you do a separate PR just for this change?

**Review by humphd — CHANGES_REQUESTED** [review]
I've pulled this locally and tested, everything worked for me.

We should use `pnpm` and set its `cache` in our workflows, see https://pnpm.io/continuous-integration#github-actions.

We need to update all of our docs, which will refer to npm.  Managing the transition to `pnpm` from `npm` is going to take education and docs.

File: .husky/pre-commit
Comment (humphd) [inline .husky/pre-commit:4]: According to https://pnpm.io/pnpm-cli#commands we don't need `run`

File: package.json
Comment (humphd) [inline package.json:10]: Same throughout this file, we can skip run i think.

File: test/test-web-content/manual-auth/index.html
Comment (humphd) [inline test/test-web-content/manual-auth/index.html:23]: Drop `run`

**Review by humphd — CHANGES_REQUESTED** [review]

File: .prettierignore
Comment (menghif) [reply .prettierignore:13]: Yes, I'll remove it from this PR. I originally added this because two files were automatically created in `src/web/public`: `sw.js` and `workbox-0ba03b8b.js` (git ignores both).
I don't know how that happened. I discovered them running `pnpm prettier-check`

**Comment by menghif** [comment]
When I run eslint, I get this message:

```console
> @senecacdot/telescope@2.3.0 eslint /Users/francesco/telescope
> eslint --config .eslintrc.js "**/*.{jsx,tsx,ts,js}"

=============

WARNING: You are currently running a version of TypeScript which is not officially supported by @typescript-eslint/typescript-estree.

You may find that it works just fine, or you may not.

SUPPORTED TYPESCRIPT VERSIONS: >=3.3.1 <4.3.0

YOUR TYPESCRIPT VERSION: 4.5.2

Please only submit bug reports when using the officially supported version.

=============
```

followed by:

**✖ 974 problems (974 errors, 0 warnings)**

**Comment by humphd** [comment]
@DukeManh was just updating eslint, do you need to rebase?

**Comment by menghif** [comment]
> @DukeManh was just updating eslint, do you need to rebase?

I already rebased this morning to get @DukeManh's changes. This is a problem related to pnpm

**Comment by humphd** [comment]
I wonder if we should be updating https://github.com/Seneca-CDOT/telescope/blob/master/package.json#L101 to https://github.com/typescript-eslint/typescript-eslint/releases/tag/v5.5.0?

**Comment by humphd** [comment]
I wonder if this is some latent dev dependency bug, where pnpm is not doing the same thing as npm with respect to pulling in peer deps.

**Comment by humphd** [comment]
What errors do you have?  What didn't it import?

One thing we might be hitting here is how we structured our dependencies using npm.  We started out with a single project, what is now `src/backend`.  When that was the case, we put everything in `package.json` in the root.  As we grew other sub-projects, we started adding other `package.json` files, but tried to put common dependencies in the root `package.json` file.

With `pnpm` this isn't necessary any more.  I think we should creating a `package.json` for `src/backend` and have the root `package.json` only contain things like jest, eslint, and other dev stuff.  We could also move the root `Dockerifile` to `src/backend`.  Doing so might fix some of this.  cc @manekenpix

**Comment by menghif** [comment]
Sorry looks like the problem was due to the fact that I removed `npm-run-all` from the dependencies.

I thought that `npm-run-all` was only used here:
https://github.com/Seneca-CDOT/telescope/blob/663b3c1da396ab080d03566eedd83532e4b21452/package.json#L27
and here:
https://github.com/Seneca-CDOT/telescope/blob/663b3c1da396ab080d03566eedd83532e4b21452/package.json#L40

Is it being used anywhere else?

---

> With pnpm this isn't necessary any more. I think we should creating a package.json for src/backend and have the root package.json only contain things like jest, eslint, and other dev stuff. We could also move the root Dockerifile to src/backend. Doing so might fix some of this.

Yes this makes sense, I will work on it.

**Comment by humphd** [comment]
We call it via `run-s`, see https://github.com/Seneca-CDOT/telescope/search?q=run-s

File: package.json
Comment (humphd) [inline package.json:27]: My concern about this change is that I think we had people on Windows who couldn't use `&&` in their shells.  Is there a reason you can't use `run-s`?

File: test/test-web-content/manual-auth/index.html
Comment (humphd) [reply test/test-web-content/manual-auth/index.html:23]: This still needs to happen.

**Review by humphd — CHANGES_REQUESTED** [review]

File: package.json
Comment (menghif) [reply package.json:27]: I'll change it back to `run-s` since we are keeping that dependency.

File: package.json
Comment (menghif) [inline package.json:30]: Removed because pnpm doesn't run pre hooks: https://pnpm.io/cli/run#differences-with-npm-run

**Comment by menghif** [comment]
@humphd I didn't know how to set up the cache in GitHub Actions so I took from your old [PR](https://github.com/Seneca-CDOT/telescope/pull/2213). I am not able to push any changes related to workflows from my computer so I am doing it via the GitHub website.

**Review by humphd — CHANGES_REQUESTED** [review]

**Comment by humphd** [comment]
The cache is working (see Actions run here)

**Comment by humphd** [comment]
This took ~11m in CI vs. 15-30m it usually takes, so already a win.

**Review by humphd — CHANGES_REQUESTED** [review]
We really need the docs updated, so we don't cause chaos when we merge this.

File: .github/workflows/node-js-ci.yml
Comment (menghif) [reply .github/workflows/node-js-ci.yml:23]: Yes! I followed the docs: https://pnpm.io/continuous-integration#github-actions

**Review by humphd — COMMENTED** [review]
This is epic.  Ping me when the lint error is fixed and I'll do final review

File: .npmrc
Comment (humphd) [inline .npmrc:4]: With this change, we might be able to switch to using `^x.y.z` style deps in `package.json` moving forward.

**Review by humphd — APPROVED** [review]
OK, I did a test.  Using [npkill](https://github.com/voidcosmos/npkill) I removed all my `node_modules` and did clean installs on `master` and with this branch.  Here's what I found:

|Package Manager |Install Time | Install Size|
|:------------------|:------------|:----------|
| `npm`                   |6m 30s        | 1.29G       | 
|  `pnpm`                | 1m 8s (!)     | 0.68G      |   

This is...wonderful.

I assume we're going to have some growing pains as we make this switch.  All devs have to be taught how to use `pnpm`, and I want to make sure it works in our deployment (cc @manekenpix), but in theory, this is worth trying.

**Comment by humphd** [comment]
@menghif can you help me draft a message to send to all the students by email before we merge this?  I'd like to cover:

- what this PR did (what it changes)
- how to adjust to use `pnpm`
- basics of how to use `pnpm`
- link(s) to where to go to get more details

You can either write it and send to me, or better yet, blog it and I'll send a link to that.

I want to make sure the transition to `pnpm` goes smoothly for our community.

**Comment by menghif** [comment]
> @menghif can you help me draft a message to send to all the students by email before we merge this? I'd like to cover:
> 
> * what this PR did (what it changes)
> * how to adjust to use `pnpm`
> * basics of how to use `pnpm`
> * link(s) to where to go to get more details
> 
> You can either write it and send to me, or better yet, blog it and I'll send a link to that.
> 
> I want to make sure the transition to `pnpm` goes smoothly for our community.

I will do this for my next blog.

**Review by DukeManh — COMMENTED** [review]
This is revolutionary! Pnpm is soooo fast.

On Gitpod though, when open a clean workspace, `pnpm` is not installed because global packages don't persist. If we want to have `pnpm` installed every time, we'd have to configure the starting point workspace Dockerfile https://www.gitpod.io/docs/config-docker#configure-a-custom-dockerfile

**Comment by humphd** [comment]
We better do the gitpod fix before we land this.  Can someone take that on in an other PR?  Could be someone else (@dhillonks are you interested?)

**Comment by menghif** [comment]
It's a quick fix. I can push it to this branch

**Review by humphd — APPROVED** [review]
Approved, pending blog post we can send to all students.

File: .gitpod.Dockerfile
Comment (DukeManh) [inline .gitpod.Dockerfile:1]: We should use a lighter workspace eventually to speed up the prebuilt process. The current `workspace/full` supports a ton of tools we don't need.

**Review by DukeManh — DISMISSED** [review]
Nicely done.

**Review by dhillonks — DISMISSED** [review]

File: .gitpod.Dockerfile
Comment (humphd) [reply .gitpod.Dockerfile:1]: Can one of you file an issue on this for follow-up?  We'll never remember.

File: docs/environment-setup.md
Comment (humphd) [inline docs/environment-setup.md:309]: Use `npx npkill` instead, which will work on all platforms.

**Review by humphd — CHANGES_REQUESTED** [review]

**Comment by humphd** [comment]
@manekenpix will landing this break deployment, since it requires a global install of pnpm?  If so, we should co-ordinate a time to do that on staging/prod.

**Comment by humphd** [comment]
I suspect we need to fix https://github.com/Seneca-CDOT/telescope/issues/2563 before we can land this.

File: .gitpod.Dockerfile
Comment (menghif) [reply .gitpod.Dockerfile:1]: Filed: #2564

**Comment by manekenpix** [comment]
@humphd these changes don't affect `staging` and `production` deployment, so I think they're safe to merge once the PR passes all the tests.

**Comment by humphd** [comment]
What's our plan to move this forward?

**Comment by menghif** [comment]
I'm hoping that #2563 will fix this. I'll work on it.

**Comment by menghif** [comment]
I rebased, update `pnpm-lock.yaml` and now CI works!

**Review by humphd — APPROVED** [review]
🥳 Really glad you stuck with this.

**Review by manekenpix — APPROVED** [review]

**Review by HyperTHD — APPROVED** [review]

**Comment by humphd** [comment]
@menghif can you get that pnpm post up and link it here so I can send to all students today?

**Comment by menghif** [comment]
> @menghif can you get that pnpm post up and link it here so I can send to all students today?

https://dev.to/menghif/transition-to-pnpm-4alf

**Pull request merged**
